Source/DafnyCore/Resolver.cs

```diff
@@ -9941,6 +9944,51 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
       return true;
     }
 
+    bool CheckIfEqualityIsDefinitelyNotSupported(IndDatatypeDecl dt, Graph<IndDatatypeDecl/*!*/>/*!*/ dependencies) {
+
+      var scc = dependencies.GetSCC(dt);
+
+      foreach (var ctor in dt.Ctors) {
+
+        if (ctor.IsGhost) {
+          return true;
+        }
+
+        foreach (var arg in ctor.Formals) {
+          var anotherIndDt = arg.Type.AsIndDatatype;
+          if (arg.IsGhost ||
+              (anotherIndDt != null && anotherIndDt.EqualitySupport == IndDatatypeDecl.ES.Never) ||
+              arg.Type.IsCoDatatype ||
+              arg.Type.IsArrowType ||
+              (arg.Type.IsOpaqueType && !arg.Type.SupportsEquality)) {
+            return true;
+          }
+
+          if (anotherIndDt != null && !scc.Contains(anotherIndDt)) {
+            CheckIfEqualityIsDefinitelyNotSupported(anotherIndDt, dependencies);
+          }
+
+          foreach (var type in arg.Type.TypeArgs) {
+            var anotherIndDt_arg = type.AsIndDatatype;
+            if ((anotherIndDt_arg != null && anotherIndDt_arg.EqualitySupport == IndDatatypeDecl.ES.Never) ||
+                type.IsCoDatatype ||
+                type.IsArrowType ||
+                (type.IsOpaqueType && !arg.Type.SupportsEquality)) {
```

Comment:
This code seems duplicated from above, let's refactor it?

Reply:
Agreed.